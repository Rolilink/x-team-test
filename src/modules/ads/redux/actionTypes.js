export const ADD_AD = 'x-team/discount-ascii-warehouse/ads/ADD';
export const CLEAR_ADS = 'x-team/discount-ascii-warehouse/ads/CLEAR';
