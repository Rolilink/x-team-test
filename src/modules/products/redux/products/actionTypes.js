import { PENDING, FULFILLED, REJECTED } from 'redux-promise-middleware';

export const FETCH_PRODUCTS = 'x-team/discount-ascii-warehouse/products/FETCH';
export const FETCH_PRODUCTS_SUCCESS = `${FETCH_PRODUCTS}_${FULFILLED}`;
export const FETCH_PRODUCTS_ERROR = `${FETCH_PRODUCTS}_${REJECTED}`;
export const FETCH_PRODUCTS_LOADING = `${FETCH_PRODUCTS}_${PENDING}`;
