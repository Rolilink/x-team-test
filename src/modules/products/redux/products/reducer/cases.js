import { cases as networkCases } from '../../../../../libs/redux-addons/redux-networking';
import { cases as paginationCases } from '../../../../../libs/redux-addons/redux-pagination';
import { createCase } from '../../../../../libs/redux-addons/redux-cases';

export const addProductsToProductsById = (state, { payload }) => ({
  ...state,
  productsById: payload.products.reduce(
    (productsById, product) => ({
      ...productsById,
      [product.id]: product,
    }),
    state.productsById,
  ),
});

const onProductFetchSuccess = createCase([
  networkCases.setSuccess,
  addProductsToProductsById,
  paginationCases.setPage,
]);

const maybeOnFetchSuccess = (state, action) => {
  const stateSort = Object.keys(state.sort)[0];
  const actionSort = Object.keys(action.payload.sort)[0];

  if (stateSort !== actionSort) {
    return state;
  }

  return onProductFetchSuccess(state, action);
};

export const onFetchSuccess = maybeOnFetchSuccess;
