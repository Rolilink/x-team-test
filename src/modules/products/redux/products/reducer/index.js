import withPagination from '../../../../../libs/redux-addons/redux-pagination';
import withSorting from '../../../../../libs/redux-addons/redux-sort';
import withNetworking from '../../../../../libs/redux-addons/redux-networking';
import makeReducer from '../../../../../libs/redux-addons/redux-make-reducer';
import composeEnhancers from '../../../../../libs/composeEnhancers';

const productsStateFactory = ({ productsById = {} } = {}) => ({ productsById });


const productsCases = { };

export default (initialState = {}) => (
  makeReducer(
    productsCases,
    productsStateFactory(initialState),
    composeEnhancers(
      [
        withPagination,
        withSorting,
        withNetworking,
      ],
      [
        initialState.pagination,
        initialState.sort,
        initialState.network,
      ],
    ),
  )
);
