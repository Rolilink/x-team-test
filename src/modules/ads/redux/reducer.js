import { handleActions } from 'redux-actions';
import { ADD_AD, CLEAR_ADS } from './actionTypes';
import { addAd, clearAds } from './cases';
import stateFactory from './factories';

export default handleActions(
  {
    [ADD_AD]: addAd,
    [CLEAR_ADS]: clearAds,
  },
  stateFactory(),
);
