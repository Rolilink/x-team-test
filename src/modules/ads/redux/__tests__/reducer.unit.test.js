import reducer from '../reducer';
import { ADD_AD, CLEAR_ADS } from '../actionTypes';

const state = {
  ads: [1, 2],
};

describe('Ads.Redux.Reducer', () => {
  it(
    'should add an ad to the ads array when dispatching an ADD_AD action',
    () => {
      const action = { type: ADD_AD, payload: { ad: 3 } };
      const updatedState = reducer(state, action);

      expect(updatedState.ads).toMatchSnapshot();
    },
  );
  it(
    'should clear the ads array when dispatching an CLEAR_ADS action',
    () => {
      const action = { type: CLEAR_ADS };
      const updatedState = reducer(state, action);

      expect(updatedState.ads).toMatchSnapshot();
    },
  );
});
