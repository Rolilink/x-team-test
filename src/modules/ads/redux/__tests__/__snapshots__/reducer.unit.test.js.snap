// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ads.Redux.Reducer should add an ad to the ads array when dispatching an ADD_AD action 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`Ads.Redux.Reducer should clear the ads array when dispatching an CLEAR_ADS action 1`] = `Array []`;
